### PR TITLE
#18397 fixes ContentletToMapTransformer bottleneck

### LINKED ISSUE #18397: Bottleneck in Field.fields(clazz)
Benchmarking this api call: `http://localhost:8080/api/v1/page/render/index?language_id=1`

```
ab -H "Authorization:Bearer $TOK" -n 10000 -c 10 http://localhost:8080/api/v1/page/json/index
```

It looks like we have a bottleneck in this code in `ContentType.java`.  The asSubclass is sloooow.  We use subclass b/c the immutables are subclasses.

```java
  @JsonIgnore
  @Value.Lazy
  public List<Field> fields(final Class<? extends Field> clazz) {
    return this.fields()
    .stream()
    .filter(field -> Try.of(()->field.getClass().asSubclass(clazz)!=null).getOrElse(false))
    .collect(Collectors.toList());
  }
```
Screen Shot 2020-04-29 at 5 50 58 PM

---

Our performance goes up like 500% If I replace this with a simple string comparison
```java
  @JsonIgnore
  @Value.Lazy
  public List<Field> fields(final Class<? extends Field> clazz) {
    final String clazzName = clazz.getName().replace(".Immutable",".");
    return this.fields()
       .stream()
       .filter(field -> 
       field.getClass().getName().replace(".Immutable",".").equals(clazzName) 
    )
    .collect(Collectors.toList());
  }
```

### PR DESCRIPTION
**Pull request by wezell**
Fixes #18397

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        
+    }
+	
+    
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:58]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        assert(!type.fields(ConstantField.class).isEmpty());
+        
+        
+        List<Field> constantsNewWay = type.fields(ConstantField.class);
+        
+        List<Field> constantsOldWay = oldWayOfGettingFields(type, ConstantField.class);
+        
+        
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert(constantsNewWay.size() >0);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:68]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
+        assert(!constants.isEmpty());
+        assert(!immutableConstants.isEmpty());
+
+        
+        
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert(constants.size() == immutableConstants.size());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:47]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+        
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
+        
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
+        assert(!constants.isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:41]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        assert(!immutableConstants.isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:42]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        
+        
+        List<Field> constantsNewWay = type.fields(ConstantField.class);
+        
+        List<Field> constantsOldWay = oldWayOfGettingFields(type, ConstantField.class);
+        
+        
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert(constantsNewWay.size() >0);
+        assert(constantsOldWay.size() >0);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:69]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	
+    @Test
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+        
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
+        
+        List<Field> constants = type.fields(ConstantField.class);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:39]: Codacy Issue found: [Local variable 'constants' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    
+    
+	@Test
+	public void test_that_fields_of_type_are_returned() throws Exception {
+	    ContentType type = TestDataUtils.getFormLikeContentType();
+	    
+	    assert(!type.fields().isEmpty());
+	    assert(type.fields().size()==type.fieldMap().size());
+	    assert(!type.fields(ConstantField.class).isEmpty());
+	    assert(type.fields(ConstantField.class).size()==2);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:26]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
+        assert(!type.fields(ConstantField.class).isEmpty());
+        
+        
+        List<Field> constantsNewWay = type.fields(ConstantField.class);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:62]: Codacy Issue found: [Local variable 'constantsNewWay' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import com.dotcms.datagen.TestDataUtils;
+import io.vavr.control.Try;
+
+public class ContentTypeTest extends ContentTypeBaseTest {
+
+
+    
+    
+	@Test
+	public void test_that_fields_of_type_are_returned() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:20]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
+        assert(!type.fields(ConstantField.class).isEmpty());
+        
+        
+        List<Field> constantsNewWay = type.fields(ConstantField.class);
+        
+        List<Field> constantsOldWay = oldWayOfGettingFields(type, ConstantField.class);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:64]: Codacy Issue found: [Local variable 'constantsOldWay' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    }
+	
+    
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
+        assert(!type.fields(ConstantField.class).isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:59]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        
+        
+        
+    }
+	
+	
+	
+	
+
+	  private List<Field> oldWayOfGettingFields(ContentType type, final Class<? extends Field> clazz) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:80]: Codacy Issue found: [Parameter 'type' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:69]: Codacy Issue found: [Substitute calls to size() == 0 (or size() != 0, size() > 0, size() < 1) with calls to isEmpty()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert(constants.size() == immutableConstants.size());
+        assertEquals(constants,immutableConstants);
+        
+    }
+	
+    
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getFormLikeContentType();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:55]: Codacy Issue found: [Local variable 'type' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:68]: Codacy Issue found: [Substitute calls to size() == 0 (or size() != 0, size() > 0, size() < 1) with calls to isEmpty()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        
+        assert(!type.fields().isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:57]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+        
+        ContentType type = TestDataUtils.getFormLikeContentType();
+        
+        assert(!type.fields().isEmpty());
+        assert(type.fields().size()==type.fieldMap().size());
+        
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:40]: Codacy Issue found: [Local variable 'immutableConstants' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+
+
+        newWayfields = type.fields(TextField.class);
+
+        oldWayFields = oldWayOfGettingFields(type, TextField.class);
+
+
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert (oldWayFields.size() > 0);
+        assert (newWayfields.size() > 0);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:102]: Codacy Issue found: [Substitute calls to size() == 0 (or size() != 0, size() > 0, size() < 1) with calls to isEmpty()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * This tests that the changing of the code in the ContentType to return fields works the same as
+     * the old way we used to do it
+     * 
+     * @throws Exception
+     */
+
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getWidgetLikeContentType();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:77]: Codacy Issue found: [Local variable 'type' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+    @Test
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+
+        ContentType type = TestDataUtils.getFormLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
+
+        List<Field> constants = type.fields(ConstantField.class);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:53]: Codacy Issue found: [Local variable 'constants' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+     */
+
+    @Test
+    public void test_that_fields_of_type_are_returned() throws Exception {
+        ContentType type = TestDataUtils.getFormLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
+        assert (!type.fields(ConstantField.class).isEmpty());
+        assert (type.fields(ConstantField.class).size() == 2);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:34]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+
+        ContentType type = TestDataUtils.getFormLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
+
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
+        assert (!constants.isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:55]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+     * the old way we used to do it
+     * 
+     * @throws Exception
+     */
+
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getWidgetLikeContentType();
+
+        assert (!type.fields().isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:79]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java (hunk 17 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:102]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+
+    /**
+     * This tests that calling type.fields() returns valid fields and that calling type.fields(Class
+     * clazz) returns a subset of those fields
+     * 
+     * @throws Exception
+     */
+
+    @Test
+    public void test_that_fields_of_type_are_returned() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:28]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+        assertEquals(oldWayFields, newWayfields);
+
+
+        newWayfields = type.fields(TextField.class);
+
+        oldWayFields = oldWayOfGettingFields(type, TextField.class);
+
+
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert (oldWayFields.size() > 0);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:101]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+     * This tests that calling type.fields(TextField.class) returns the same subset of fields as does
+     * calling type.fields(ImmutableTextField.class)
+     * 
+     * @throws Exception
+     */
+
+    @Test
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+
+        ContentType type = TestDataUtils.getFormLikeContentType();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:48]: Codacy Issue found: [Local variable 'type' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+     * @throws Exception
+     */
+
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getWidgetLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
+        assert (!type.fields(ConstantField.class).isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:81]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
+        assert (!constants.isEmpty());
+        assert (!immutableConstants.isEmpty());
+
+
+
+        // make sure that getting the field by Type or by its immutable type works the same
+        assert (constants.size() == immutableConstants.size());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:61]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+
+        ContentType type = TestDataUtils.getFormLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
+
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
+        assert (!constants.isEmpty());
+        assert (!immutableConstants.isEmpty());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:56]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 29]
```diff
@@ -0,0 +1,122 @@
...
+    @Test
+    public void test_that_immutable_fields_work_the_same_as_normal_fields_returned() throws Exception {
+
+        ContentType type = TestDataUtils.getFormLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
+
+        List<Field> constants = type.fields(ConstantField.class);
+        List<Field> immutableConstants = type.fields(ImmutableConstantField.class);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:54]: Codacy Issue found: [Local variable 'immutableConstants' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+     * 
+     * @throws Exception
+     */
+
+    @Test
+    public void test_that_it_works_like_the_old_way() throws Exception {
+        ContentType type = TestDataUtils.getWidgetLikeContentType();
+
+        assert (!type.fields().isEmpty());
+        assert (type.fields().size() == type.fieldMap().size());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeTest.java:80]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5449504)

**Review by dsilvam — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Review by jdotcms — COMMENTED** [review]
good

**Pull request merged**